### PR TITLE
Moving Shopping Cart to new design

### PR DESCRIPTION
**Pull request by AntonioFalcaoJr**
* **Please check if the PR fulfills these requirements**
- [ ] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

* **What is the current behavior?** (You can also link to an open issue here)

* **What is the new behavior (if this is a feature change)?**

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

* **Other information**:

### CODE REVIEW

File: src/Contracts/Services/ShoppingCart/Query.cs [hunk 1]
```diff
@@ -1,18 +1,43 @@
...
 
 namespace Contracts.Services.ShoppingCart;
 
 public static class Query
 {
-    public record GetShoppingCart(Guid CartId) : Message, IQuery;
+    public record GetShoppingCartDetails(Guid CartId) : IQuery
+    {
+        public static implicit operator GetShoppingCartDetails(Protobuf.GetShoppingCartDetailsRequest request)
+            => new(new Guid(request.CartId));
```
Comment (MarcusViniciusSS) [inline src/Contracts/Services/ShoppingCart/Query.cs:11]: Here is possible use this strategy.

```suggestion
            => new(new(request.CartId));
```

File: src/Services/Account/Query/Infrastructure.Projections/ProjectionGateway.cs [hunk 2]
```diff
@@ -32,8 +32,12 @@ public ProjectionGateway(IMongoDbContext context)
     public Task<IPagedResult<TProjection>?> ListAsync(Paging paging, CancellationToken cancellationToken)
         => PagedResult<TProjection>.CreateAsync(paging, _collection.AsQueryable(), cancellationToken)!;
 
-    public Task InsertAsync(TProjection projection, CancellationToken cancellationToken)
-        => _collection.InsertOneAsync(projection, cancellationToken: cancellationToken);
+    public Task ReplaceAsync(TProjection replacement, CancellationToken cancellationToken)
```
Comment (MarcusViniciusSS) [inline src/Services/Account/Query/Infrastructure.Projections/ProjectionGateway.cs:35]: This method is equal the UpsertAsync , my suggestion is remove it.

File: src/Services/Catalog/Query/Application/Abstractions/IProjectionGateway.cs [hunk 3]
```diff
@@ -11,7 +11,7 @@ public interface IProjectionGateway<TProjection>
     Task<TProjection?> GetAsync<TId>(TId id, CancellationToken cancellationToken) where TId : struct;
     Task<IPagedResult<TProjection>?> ListAsync(Paging paging, Expression<Func<TProjection, bool>> predicate, CancellationToken cancellationToken);
     Task<IPagedResult<TProjection>?> ListAsync(Paging paging, CancellationToken cancellationToken);
-    Task InsertAsync(TProjection projection, CancellationToken cancellationToken);
+    Task ReplaceAsync(TProjection replacement, CancellationToken cancellationToken);
```
Comment (MarcusViniciusSS) [inline src/Services/Catalog/Query/Application/Abstractions/IProjectionGateway.cs:14]: remove here is equal Upsert.

File: src/Services/Catalog/Query/Infrastructure.Projections/ProjectionGateway.cs [hunk 4]
```diff
@@ -32,8 +32,12 @@ public ProjectionGateway(IMongoDbContext context)
     public Task<IPagedResult<TProjection>?> ListAsync(Paging paging, CancellationToken cancellationToken)
         => PagedResult<TProjection>.CreateAsync(paging, _collection.AsQueryable(), cancellationToken)!;
 
-    public Task InsertAsync(TProjection projection, CancellationToken cancellationToken)
-        => _collection.InsertOneAsync(projection, cancellationToken: cancellationToken);
+    public Task ReplaceAsync(TProjection replacement, CancellationToken cancellationToken)
```
Comment (MarcusViniciusSS) [inline src/Services/Catalog/Query/Infrastructure.Projections/ProjectionGateway.cs:35]: Remove it.

File: src/Services/Communication/Query/Application/Abstractions/IProjectionGateway.cs [hunk 5]
```diff
@@ -11,7 +11,7 @@ public interface IProjectionGateway<TProjection>
     Task<TProjection?> GetAsync<TId>(TId id, CancellationToken cancellationToken) where TId : struct;
     Task<IPagedResult<TProjection>?> ListAsync(Paging paging, Expression<Func<TProjection, bool>> predicate, CancellationToken cancellationToken);
     Task<IPagedResult<TProjection>?> ListAsync(Paging paging, CancellationToken cancellationToken);
-    Task InsertAsync(TProjection projection, CancellationToken cancellationToken);
+    Task ReplaceAsync(TProjection replacement, CancellationToken cancellationToken);
```
Comment (MarcusViniciusSS) [inline src/Services/Communication/Query/Application/Abstractions/IProjectionGateway.cs:14]: remove it

File: src/Services/Order/Command/Domain/ValueObjects/PaymentOptions/PayPals/PayPalValidator.cs [hunk 6]
```diff
@@ -10,8 +10,7 @@ public PayPalValidator()
             .NotNull()
             .NotEmpty();
 
-        RuleFor(card => card.UserName)
-            .NotNull()
-            .NotEmpty();
+        RuleFor(card => card.Email)
```
Comment (MarcusViniciusSS) [inline src/Services/Order/Command/Domain/ValueObjects/PaymentOptions/PayPals/PayPalValidator.cs:13]: I suggestion use **.NotEmpty()** together if this field is required,

File: src/Services/Order/Query/Application/Abstractions/IProjectionGateway.cs [hunk 7]
```diff
@@ -11,7 +11,7 @@ public interface IProjectionGateway<TProjection>
     Task<TProjection?> GetAsync<TId>(TId id, CancellationToken cancellationToken) where TId : struct;
     Task<IPagedResult<TProjection>?> ListAsync(Paging paging, Expression<Func<TProjection, bool>> predicate, CancellationToken cancellationToken);
     Task<IPagedResult<TProjection>?> ListAsync(Paging paging, CancellationToken cancellationToken);
-    Task InsertAsync(TProjection projection, CancellationToken cancellationToken);
+    Task ReplaceAsync(TProjection replacement, CancellationToken cancellationToken);
```
Comment (MarcusViniciusSS) [inline src/Services/Order/Query/Application/Abstractions/IProjectionGateway.cs:14]: Duplicate

File: src/Services/Order/Query/Infrastructure.Projections/ProjectionGateway.cs [hunk 8]
```diff
@@ -32,8 +32,12 @@ public ProjectionGateway(IMongoDbContext context)
     public Task<IPagedResult<TProjection>?> ListAsync(Paging paging, CancellationToken cancellationToken)
         => PagedResult<TProjection>.CreateAsync(paging, _collection.AsQueryable(), cancellationToken)!;
 
-    public Task InsertAsync(TProjection projection, CancellationToken cancellationToken)
-        => _collection.InsertOneAsync(projection, cancellationToken: cancellationToken);
+    public Task ReplaceAsync(TProjection replacement, CancellationToken cancellationToken)
```
Comment (MarcusViniciusSS) [inline src/Services/Order/Query/Infrastructure.Projections/ProjectionGateway.cs:35]: this code is equal upsertAsync.

File: src/Services/ShoppingCart/Command/Infrastructure.MessageBus/DependencyInjection/Extensions/RabbitMqBusFactoryConfiguratorExtensions.cs [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+    public static void ConfigureEventReceiveEndpoints(this IRabbitMqBusFactoryConfigurator cfg, IRegistrationContext context)
+    {
+
+    }
+
+    private static void ConfigureEventReceiveEndpoint<TConsumer, TEvent>(this IRabbitMqBusFactoryConfigurator bus, IRegistrationContext context)
+        where TConsumer : class, IConsumer
+        where TEvent : class, IEvent
+        => bus.ReceiveEndpoint(
+            queueName: $"catalog.command-stack.{typeof(TConsumer).ToKebabCaseString()}.{typeof(TEvent).ToKebabCaseString()}",
```
Comment (MarcusViniciusSS) [inline src/Services/ShoppingCart/Command/Infrastructure.MessageBus/DependencyInjection/Extensions/RabbitMqBusFactoryConfiguratorExtensions.cs:17]:
```suggestion
            queueName: $"shopping-cart.command-stack.{typeof(TConsumer).ToKebabCaseString()}.{typeof(TEvent).ToKebabCaseString()}",
```

File: src/Services/ShoppingCart/Command/Infrastructure.MessageBus/PipeFilters/BusinessValidatorFilter.cs [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+    where T : class
+{
+    public async Task Send(ExceptionConsumeContext<T> context, IPipe<ExceptionConsumeContext<T>> next)
+    {
+        if (context.Exception is ValidationException exception)
+        {
+            Log.Error("Business validation errors: {Errors}", exception.Errors);
+
+            await context.Send(
+                destinationAddress: new($"queue:catalog.{KebabCaseEndpointNameFormatter.Instance.SanitizeName(typeof(T).Name)}.business-error"),
```
Comment (MarcusViniciusSS) [inline src/Services/ShoppingCart/Command/Infrastructure.MessageBus/PipeFilters/BusinessValidatorFilter.cs:18]:
```suggestion
                destinationAddress: new($"queue:shopping-cart.{KebabCaseEndpointNameFormatter.Instance.SanitizeName(typeof(T).Name)}.business-error"),
```

File: src/Services/ShoppingCart/Command/Infrastructure.MessageBus/PipeFilters/ContractValidatorFilter.cs [hunk 11]
```diff
@@ -34,7 +34,7 @@ public async Task Send(ConsumeContext<T> context, IPipe<ConsumeContext<T>> next)
         Log.Error("Contract validation errors: {Errors}", validationResult.Errors);
 
         await context.Send(
-            destinationAddress: new($"queue:shopping-cart.{KebabCaseEndpointNameFormatter.Instance.SanitizeName(typeof(T).Name)}.contract-errors"),
+            destinationAddress: new($"queue:catalog.{KebabCaseEndpointNameFormatter.Instance.SanitizeName(typeof(T).Name)}.contract-errors"),
```
Comment (MarcusViniciusSS) [inline src/Services/ShoppingCart/Command/Infrastructure.MessageBus/PipeFilters/ContractValidatorFilter.cs:37]:
```suggestion
            destinationAddress: new($"queue:shopping-cart.{KebabCaseEndpointNameFormatter.Instance.SanitizeName(typeof(T).Name)}.contract-errors"),
```

File: src/Services/ShoppingCart/Command/WorkerService/appsettings.json [hunk 12]
```diff
@@ -22,7 +22,7 @@
     "SnapshotInterval": 5
   },
   "QuartzOptions": {
-    "quartz.scheduler.instanceName": "ShoppingCart",
+    "quartz.scheduler.instanceName": "Catalog",
```
Comment (MarcusViniciusSS) [inline src/Services/ShoppingCart/Command/WorkerService/appsettings.json:25]:
```suggestion
    "quartz.scheduler.instanceName": "ShoppingCart",
```

File: src/Services/ShoppingCart/Query/GrpcService/Properties/launchSettings.json [hunk 13]
```diff
@@ -0,0 +1,13 @@
+﻿{
+  "profiles": {
+    "Catalog.GrpcService": {
+      "commandName": "Project",
+      "dotnetRunMessages": true,
+      "launchBrowser": false,
+      "applicationUrl": "https://localhost:7218;http://localhost:5133",
```
Comment (MarcusViniciusSS) [inline src/Services/ShoppingCart/Query/GrpcService/Properties/launchSettings.json]: change the port of localhost for **last port + 1**

File: src/Services/ShoppingCart/Query/GrpcService/Properties/launchSettings.json [hunk 14]
```diff
@@ -0,0 +1,13 @@
+﻿{
+  "profiles": {
+    "Catalog.GrpcService": {
```
Comment (MarcusViniciusSS) [inline src/Services/ShoppingCart/Query/GrpcService/Properties/launchSettings.json]: ShoppingCart.G
```suggestion
    "ShoppingCart.GrpcService": {
```

File: src/Services/ShoppingCart/Query/GrpcService/appsettings.Development.json [hunk 15]
```diff
@@ -0,0 +1,8 @@
+{
+  "ConnectionStrings": {
+    "Projections": "mongodb://mongoadmin:secret@127.0.0.1:27017/CatalogProjections/?authSource=admin"
```
Comment (MarcusViniciusSS) [inline src/Services/ShoppingCart/Query/GrpcService/appsettings.Development.json:3]:
```suggestion
    "Projections": "mongodb://mongoadmin:secret@127.0.0.1:27017/ShoppingCartProjections/?authSource=admin"
```

File: src/Services/ShoppingCart/Query/GrpcService/appsettings.Production.json [hunk 16]
```diff
@@ -0,0 +1,8 @@
+{
+  "ConnectionStrings": {
+    "Projections": "mongodb://mongoadmin:secret@mongodb/CatalogProjections/?authSource=admin"
```
Comment (MarcusViniciusSS) [inline src/Services/ShoppingCart/Query/GrpcService/appsettings.Production.json:3]:
```suggestion
    "Projections": "mongodb://mongoadmin:secret@mongodb/ShoppingCartProjections/?authSource=admin"
```

File: src/Services/ShoppingCart/Query/GrpcService/appsettings.Staging.json [hunk 17]
```diff
@@ -0,0 +1,8 @@
+{
+  "ConnectionStrings": {
+    "Projections": "mongodb://mongoadmin:secret@mongodb/CatalogProjections/?authSource=admin"
```
Comment (MarcusViniciusSS) [inline src/Services/ShoppingCart/Query/GrpcService/appsettings.Staging.json:3]:
```suggestion
    "Projections": "mongodb://mongoadmin:secret@mongodb/ShoppingCartProjections/?authSource=admin"
```

File: src/Services/ShoppingCart/Query/Infrastructure.EventBus/DependencyInjection/Extensions/RabbitMqBusFactoryConfiguratorExtensions.cs [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+        cfg.ConfigureEventReceiveEndpoint<ProjectCartItemListItemWhenCartChangedConsumer, DomainEvent.CartItemDecreased>(context);
+
+        cfg.ConfigureEventReceiveEndpoint<ProjectPaymentMethodListItemWhenCartChangedConsumer, DomainEvent.PaymentMethodAdded>(context);
+    }
+
+    private static void ConfigureEventReceiveEndpoint<TConsumer, TEvent>(this IRabbitMqBusFactoryConfigurator bus, IRegistrationContext context)
+        where TConsumer : class, IConsumer
+        where TEvent : class, IEvent
+        => bus.ReceiveEndpoint(
+            queueName: $"catalog.query-stack.{typeof(TConsumer).ToKebabCaseString()}.{typeof(TEvent).ToKebabCaseString()}",
```
Comment (MarcusViniciusSS) [inline src/Services/ShoppingCart/Query/Infrastructure.EventBus/DependencyInjection/Extensions/RabbitMqBusFactoryConfiguratorExtensions.cs:36]: **shopping-cart**

File: src/Services/ShoppingCart/Query/Infrastructure.EventBus/PipeFilters/ContractValidatorFilter.cs [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+        if (validationResult.IsValid)
+        {
+            await next.Send(context);
+            return;
+        }
+
+        Log.Error("Contract validation errors: {Errors}", validationResult.Errors);
+
+        await context.Send(
+            destinationAddress: new($"queue:catalog.{KebabCaseEndpointNameFormatter.Instance.SanitizeName(typeof(T).Name)}.contract-errors"),
```
Comment (MarcusViniciusSS) [inline src/Services/ShoppingCart/Query/Infrastructure.EventBus/PipeFilters/ContractValidatorFilter.cs:37]: shopping-cart

Comment (MarcusViniciusSS) [inline src/Services/ShoppingCart/Query/Infrastructure.EventBus/PipeFilters/ContractValidatorFilter.cs:37]: shopping-cart
```suggestion
            destinationAddress: new($"queue:shopping-cart.{KebabCaseEndpointNameFormatter.Instance.SanitizeName(typeof(T).Name)}.contract-errors"),
```

File: src/Web/WebAPI/APIs/Payments/PaymentApi.cs [hunk 20]
```diff
@@ -24,8 +20,8 @@ public static IVersionedEndpointRouteBuilder MapPaymentApiV2(this IVersionedEndp
     {
         var group = builder.MapGroup(BaseUrl).HasApiVersion(2);
 
-        group.MapGet("/{paymentId:guid}", (IBus bus, [NotEmpty] Guid paymentId, CancellationToken cancellationToken)
-            => ApplicationApi.GetProjectionAsync<Query.GetPayment, Projection.Payment>(bus, new(paymentId), cancellationToken));
+        // group.MapGet("/{paymentId:guid}", (IBus bus, [NotEmpty] Guid paymentId, CancellationToken cancellationToken)
```
Comment (MarcusViniciusSS) [inline src/Web/WebAPI/APIs/Payments/PaymentApi.cs:23]: ???

File: src/Web/WebAPI/APIs/Payments/PaymentApi.cs [hunk 21]
```diff
@@ -14,8 +10,8 @@ public static IVersionedEndpointRouteBuilder MapPaymentApiV1(this IVersionedEndp
     {
         var group = builder.MapGroup(BaseUrl).HasApiVersion(1);
 
-        group.MapGet("/{paymentId:guid}", (IBus bus, [NotEmpty] Guid paymentId, CancellationToken cancellationToken)
-            => ApplicationApi.GetProjectionAsync<Query.GetPayment, Projection.Payment>(bus, new(paymentId), cancellationToken));
+        // group.MapGet("/{paymentId:guid}", (IBus bus, [NotEmpty] Guid paymentId, CancellationToken cancellationToken)
```
Comment (MarcusViniciusSS) [inline src/Web/WebAPI/APIs/Payments/PaymentApi.cs:13]: ???

File: src/Web/WebAPI/APIs/Payments/PaymentApi.cs (hunk 20 above)
Comment (AntonioFalcaoJr) [reply src/Web/WebAPI/APIs/Payments/PaymentApi.cs:23]: breaking changes

File: src/Web/WebAPI/APIs/Payments/PaymentApi.cs (hunk 21 above)
Comment (AntonioFalcaoJr) [reply src/Web/WebAPI/APIs/Payments/PaymentApi.cs:13]: breaking changes

**Pull request merged**